oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoClusterListener.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.jackrabbit.oak.plugins.document.mongo;
+
+import com.mongodb.ServerAddress;
+import com.mongodb.connection.ServerConnectionState;
+import com.mongodb.connection.ServerDescription;
+import com.mongodb.event.ClusterDescriptionChangedEvent;
+import com.mongodb.event.ClusterListener;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+public class MongoClusterListener implements ClusterListener {
+
+    // Sometimes we need to wait a few seconds in case the connection was just created, the listener
+    // didn't have time to receive the description from the cluster. This latch is used to check
+    // if the connection was properly initialized.
+    private final CountDownLatch LATCH;
+    private final int LATCH_AWAIT_TIMEOUT = 15;
+    private boolean replicaSet = false;
+    private ServerAddress serverAddress;
+    private ServerAddress primaryAddress;
+
+    public MongoClusterListener() {
+        LATCH = new CountDownLatch(1);
+    }
+
+    @Nullable
+    public ServerAddress getServerAddress() {
+        waitForLatch();
```

Comment:
What would happen in case `waitForLatch` timed out and returned false?

Reply:
The idea behind the Latch mechanism is to wait a few seconds until a new connection is fully established. This only happens when the connection is really fresh. After the first `clusterDescriptionChanged` is called, the variables are filled in and the latch is no longer used.

If the latch times out, the ServerAddress returned will be null, which shouldn't be a problem as it has the `Nullable` annotation and when it's used it is checked for null. For example: https://github.com/apache/jackrabbit-oak/blob/c6ddcc55bee3de915459af01e91edad32d538f3d/oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoStatusTest.java#L207

Reply:
Following your comment, I'm adding the `Nullable` annotation to the 3 methods in [MongoConnection.java](https://github.com/apache/jackrabbit-oak/blob/fada62153fe314c144ba6b8ccc9155438fe8f2a8/oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/MongoConnection.java), so it clearly states the returned object can be null:
https://github.com/apache/jackrabbit-oak/blob/fada62153fe314c144ba6b8ccc9155438fe8f2a8/oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/MongoConnection.java#L352

And the following ones.